examples/iterative_probabilistic_kh.py

```diff
@@ -0,0 +1,182 @@
+        n_samples=num_data_points,
+        n_features=num_features,
+        centers=num_cluster_centers,
+        random_state=random_seed,
+        return_centers=True,
+        cluster_std=2,
+    )
+    return Data(jnp.asarray(x))
+
+
+def get_kernels(
+    data: Data, random_seed: int = 0
+) -> tuple[SquaredExponentialKernel, SquaredExponentialKernel]:
+    """Get a SquaredExponentialKernel based on the data using the median heuristic."""
+    num_data_points = len(data)
+    num_samples_length_scale = min(num_data_points, 1_000)
+    generator = np.random.default_rng(random_seed)
+    idx = generator.choice(num_data_points, num_samples_length_scale, replace=False)
+    length_scale = median_heuristic(data.data[idx])
+    return (
+        SquaredExponentialKernel(length_scale=length_scale),
+        SquaredExponentialKernel(
+            length_scale=length_scale,
+            output_scale=1.0 / (length_scale * jnp.sqrt(2.0 * jnp.pi)),
+        ),
+    )
+
+
+def iterative_refine_experiment(
+    data: Data,
+    coreset_size: int,
+    n_iter: int,
+    t_schedule: Optional[Array] = None,
+    seed: int = 0,
+) -> tuple[Array, Coresubset]:
+    """Perform an experiment with n_iter iterations and a given temperature schedule."""
+    random_key = jax.random.key(seed)
+    kernel, mmd_kernel = get_kernels(data)
+    mmd_metric = MMD(mmd_kernel)
+    probabilistic = True
+    if t_schedule is None:
+        t_schedule = jnp.ones(n_iter)
+        probabilistic = False
+
+    mmd_data = jnp.zeros(n_iter)  # store experiment data
+    initial_coreset = _initial_coresubset(0, coreset_size, data)
+
+    def run_exp(i, state):
+        coreset, mmd_data, key = state
+        key = jax.random.fold_in(key, i)
+
+        solver = KernelHerding(
+            coreset_size=coreset_size,
+            kernel=kernel,
+            probabilistic=probabilistic,
+            temperature=t_schedule[i],
+            random_key=key,
+        )
+        coreset, _ = solver.refine(coreset)
+        mmd_data = mmd_data.at[i].set(coreset.compute_metric(mmd_metric))
+        return coreset, mmd_data, key
+
+    coreset, mmd_data, random_key = jax.lax.fori_loop(
+        0, n_iter, run_exp, (initial_coreset, mmd_data, random_key)
+    )
+    return mmd_data, coreset
+
+
+def visualise_results(mmd_data_prob: Dict[str, Array], mmd_data_base: Array) -> None:
+    """Visualise the results of the experiment."""
+    baseline = mmd_data_base[0].item()
+    plt.plot(mmd_data_base, label="Standard KH")
+    plt.axhline(baseline, c="k", ls="--", label="Standard KH reduce")
+
+    for label, mmd_data in mmd_data_prob.items():
+        plt.plot(mmd_data, label=f"Probabilistic: {label}")
+
+    plt.grid(ls=":")
+    plt.xlabel("Refinement iteration")
+    plt.ylabel("MMD")
+    plt.legend()
+    plt.ylim([0, baseline * 3])  # zoom in on the relevant scale
+    plt.title("Iterative refinement with probabilistic selection")
+    plt.show()
+
+
+if __name__ == "__main__":
```

Comment:
```suggestion
def main() -> None:
```
followed by 
```python
if __name__ == "__main__":
    main()
```
at the bottom of the file. This minimises global variables which should help prevent errors.